### PR TITLE
Use evented dataclass for dims

### PR DESCRIPTION
**Pull request by sofroniewn**
# Description
Continues a series of PRs which should all be reviewed before this one. This makes the `dims` and eventeddata class and supersedes previous efforts here #1489, and #1703 and #1486 which tried to use an evented list for some properties too, this PR doesn't use the evented list in dims, that could come later.

It does have a breaking change which is that the `axis` parameter is no longer present on the `current_step`, `range`, or `axis_labels` events. Instead a single event is emitted whenever that one of those changes at all.

Also all the list-like properties are now tuples instead of lists, which is much more accurate as they actually can't be edited.

## Type of change

- [x] Refactor (non-breaking change which fixes an issue)

### CODE REVIEW

**Comment by sofroniewn** [comment]
This PR is now ready for review. It moves over the `Dims` to now be an `evented_dataclass`. Since we merge #1919 which removed the `Dims` object from the individual layers, the only place where `Dims` is now used is on the `ViewerModel` which is quite nice!

This PR has one breaking change which is that the axis parameter is no longer present on the current_step, range, or axis_labels events. Instead a single event is emitted whenever that one of those changes at all. I'm curious if @tlambert03 or @Czaki spot any trouble that this may cause for us. If so we might want to change.

Also all the list-like properties are now tuples instead of lists, which is much more accurate as they actually can't be edited. We could use the `evented_list` at some point, but to me that need is less pressing then getting this merged with the `evented_dataclass`.

Would be great to get feedback from @jni @tlambert03 @Czaki and any others interested in the `Dims` model!!

File: napari/components/dims.py [hunk 1]
```diff
@@ -1,265 +1,220 @@
...
-            source=self,
-            auto_connect=False,
-            current_step=None,
-            axis_labels=None,
-            ndim=None,
-            ndisplay=None,
-            order=None,
-            range=None,
-            last_used=None,
-            deprecated={"axis": "current_step", "camera": "ndisplay"},
```
Comment (Czaki) [inline napari/components/dims.py:63]: So now there is no deprecation mechanism, so `camera` is dropped?

File: napari/components/dims.py [hunk 2]
```diff
@@ -1,265 +1,220 @@
...
-    @axis_labels.setter
-    def axis_labels(self, labels):
-        if self._axis_labels == labels:
-            return
+        self._ndim = ndim
 
-        if len(labels) != self.ndim:
+    def _on_order_set(self, order):
+        if not set(order) == set(range(self.ndim)):
+            print('eeeeee')
```
Comment (Czaki) [inline napari/components/dims.py]: Removed print?

File: napari/components/dims.py [hunk 3]
```diff
@@ -1,265 +1,220 @@
...
-        if np.all(self._order == order):
-            return
-
-        if not len(order) == self.ndim:
+    def _on_axis_labels_set(self, axis_labels):
+        if not len(axis_labels) == self.ndim:
             raise ValueError(
-                f"Invalid ordering {order} for {self.ndim} dimensions"
+                f"Invalid number of axis labels {len(axis_labels)} for {self.ndim} dimensions"
             )
```
Comment (Czaki) [inline napari/components/dims.py]: I see that we heck here ndim, but in code of ndim there is no interaction with axis labels.

File: napari/components/dims.py [hunk 4]
```diff
@@ -1,265 +1,220 @@
...
-            for axis_changed in range(ndim - cur_ndim):
-                self.events.current_step(axis=axis_changed)
-        elif ndim < cur_ndim:
-            self._range = self._range[-ndim:]
-            self._current_step = self._current_step[-ndim:]
-            self._order = self._reorder_after_dim_reduction(
-                self._order[-ndim:]
+    def _on_range_set(self, range_var):
+        if not len(range_var) == self.ndim:
+            print('hasdfhsd')
```
Comment (Czaki) [inline napari/components/dims.py]: remove

**Review by Czaki — COMMENTED** [review]
Few questions

File: napari/components/dims.py (hunk 2 above)
Comment (sofroniewn) [reply napari/components/dims.py]: thanks!! removed

File: napari/components/dims.py (hunk 4 above)
Comment (sofroniewn) [reply napari/components/dims.py]: removed!

File: napari/components/dims.py (hunk 1 above)
Comment (sofroniewn) [reply napari/components/dims.py:63]: That's correct, those both were deprecated and now are being dropped.

File: napari/components/dims.py (hunk 3 above)
Comment (sofroniewn) [reply napari/components/dims.py]: The axis_labels length are adjusted inside the `ndim` set method here https://github.com/napari/napari/blob/e4c7b37a7c79602c8b1ea7d543bb3a8cfdbe1f14/napari/components/dims.py#L162-L174. Is that what you were looking for?

Comment (Czaki) [reply napari/components/dims.py]: Yes. sorry.

File: napari/components/dims.py [hunk 5]
```diff
@@ -1,265 +1,218 @@
...
-
-    @property
-    def point(self):
-        """List of float: value of each dimension."""
-        # The point value is computed from the current_step
-        point = [
-            min_val + step_size * value
-            for (min_val, max_val, step_size), value in zip(
-                self._range, self._current_step
+    def _on_ndim_set(self, ndim):
```
Comment (jni) [inline napari/components/dims.py:141]: I'm probably misunderstanding something and need to read the evented dataclass docs, but how/when does this get called?

**Review by jni — APPROVED** [review]
Great, thanks! I only left one question, which could probably be addressed in a comment in the corresponding code.

I do think long term we want to move to evented lists for the order/axis labels etc, but agree that this can come later.

File: napari/components/dims.py (hunk 5 above)
Comment (sofroniewn) [reply napari/components/dims.py:141]: These get called after the setter, see [here in the code](https://github.com/napari/napari/blob/master/napari/utils/events/dataclass.py#L174-L187) and this comment describing it. https://github.com/napari/napari/pull/1489#discussion_r460423875. I can add a comment

**Comment by sofroniewn** [comment]
I've added some comments and doc strings. Will merge after tests pass

**Pull request merged**